### PR TITLE
Use unforgeable name as mergeable channel tag

### LINKED ISSUE #3513: REV vaults (number channel) merging
## Overview

PoS (Proof of Stake contract) needs to collect validator rewards in PoS vault. If PoS vault is used directly in each block it will create conflict when merging blocks. In attempt to resolve this current implementation of PoS creates REV vault for each active validator (per validator vaults).
This is not enough when multiple blocks are created by the same validator in which case merging must reject the whole block because per validator vault will be conflicting.
https://github.com/rchain/rchain/blob/cb5ec1328132d92425935eeea478ffce148f94be/casper/src/main/resources/PoS.rhox#L55-L72

Another issue is slow refund operation in PoS which not only return unused REV to the deployer but also distribute rewards to validators by their stake.
https://github.com/rchain/rchain/blob/cb5ec1328132d92425935eeea478ffce148f94be/casper/src/main/resources/PoS.rhox#L295-L308

## Solution

As a first implementation merging number channels will be done for channels holding REV balances which means that PoS don't need per validator vaults and can directly work with PoS vault.

To resolve slow refund, distribution of rewards should be done only on epoch change or slashing and not together with refund operation which will speed up block processing time.

## Design

To implement merging of REV balances to major changes needs to be done.

- [x] 1. Modify the Rholang reducer to collect channel hashes which are used to hold REV balance in NonNegativeNumber contract.
    #3492
- [x] 2. Modify Rholang Runtime to persist data about mergeable channels (REV balances) on disk.
    #3492
- [x] 3. Modify conflict detection and deploy dependency calculation to skip mergeable channels as a special case.
    #3492
- [x] 4. Modify mergeable logic to support for merging values on mergeable channels, in this case Rholang `Int` types used as REV balances.
    #3492
- [x] #3514
  #3559
    - Rewards distribution should be done lazily only when is necessary and not on every block (epoch change, slashing).
    - PoS address should be replaced with multisig vault (the same as existing coop vault).
- [x] Create unforgeable name to tag number channels, now fixed __MERGEABLE__ name is used.
  #3577
- [x] Implement rejections for merged number channels which are negative or overflow.
  #3591
- [x] Add additional tests for number channel merging logic.

### LINKED ISSUE #3575: Implement the safe way to tag mergeable number channels
## Overview

@tgrospic introduce a new way to resolving conflicts with mergeable channels on numbers in https://github.com/rchain/rchain/pull/3492. But currently the codes is using a "__MERGEABLE__" string to 
tag these mergeable channels which is not safe at all

The issue is intended to explore a proper way to fix that.

1. replace __MERGEABLE__ with an unforgeable name
2. add config for mergeable tag to runtime

### Related issue
- https://github.com/rchain/rchain/issues/3513

### PR DESCRIPTION
**Pull request by zsluedem**
## Overview

Create an unforgeable name to tag number channels. Now the code is using fixes string to do that which is not safe at all and it is exposed to the user.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala [hunk 1]
```diff
@@ -67,7 +68,7 @@ object RholangAndScalaDispatcher {
       urnMap: Map[String, Par]
   ): (Dispatch[M, ListParWithRandom, TaggedContinuation], Reduce[M]) = {
     val emptyMergeableRef = Ref.unsafe[M, Set[Par]](Set.empty)
-
-    apply(tuplespace, dispatchTable, urnMap, emptyMergeableRef)
+    val dummyMergeableTag = Par()
```
Comment (tgrospic) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: Why to use dummy mergeable tag name here?

Comment (zsluedem) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: For some tests who don't care about mergeable tag name. Is it better to move it to test package?

File: node/src/main/scala/coop/rchain/node/revvaultexport/mainnet1/StateBalanceMain.scala [hunk 2]
```diff
@@ -59,7 +59,7 @@ object StateBalanceMain {
   // The way to get this Unforgeable name needs reporting casper to get all the concrete comm events.
   // Anyway, as long as RevVault.rho and genesis doesn't change, this value would be fixed.
   val genesisVaultMapPar: Par = GPrivate(
-    "2dea7213bab34c96d4d3271f9b9dcb0b50a242de7de2177da615d9de23106afd".unsafeHexToByteString
+    "5e8200d74ba8689013cf86f075775a18a42f1bd3a5ea122e82dfca9ba485a924".unsafeHexToByteString
```
Comment (tgrospic) [inline node/src/main/scala/coop/rchain/node/revvaultexport/mainnet1/StateBalanceMain.scala:62]: Do we have a test to check this hard coded value?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/RhoRuntime.scala [hunk 3]
```diff
@@ -478,7 +480,8 @@ object RhoRuntime {
   def createRhoEnv[F[_]: Concurrent: Parallel: _cost: Log: Metrics: Span](
       rspace: RhoISpace[F],
       mergeChs: Ref[F, Set[Par]],
-      extraSystemProcesses: Seq[Definition[F]] = Seq.empty
+      extraSystemProcesses: Seq[Definition[F]] = Seq.empty,
+      mergeableTagName: Par
```
Comment (tgrospic) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/RhoRuntime.scala]: It's more flexible to put parameters with defaults at the end.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala (hunk 1 above)
Comment (tgrospic) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: Yes, that would be better.

Also keep in mind that reducer will detect any tuple with this form `(MergeableTag, _)`, so with `Par()` as a mergeable tag this means any tuple `(Nil, _)` will be considered mergeable.
What I mean is that dummy tag must be something not possible to construct in testing Rholang code.

Comment (zsluedem) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: Good point. How about `Par(unforgeables=something)`?

File: node/src/main/scala/coop/rchain/node/revvaultexport/mainnet1/StateBalanceMain.scala (hunk 2 above)
Comment (zsluedem) [reply node/src/main/scala/coop/rchain/node/revvaultexport/mainnet1/StateBalanceMain.scala:62]: Yes. https://github.com/rchain/rchain/blob/dev/node/src/test/scala/coop/rchain/node/revvaultexport/VaultBalanceGetterTest.scala

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala (hunk 1 above)
Comment (zsluedem) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: I moved this `apply` to test package since it is only used in tests.

Comment (tgrospic) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: > Good point. How about `Par(unforgeables=something)`?

Yes, in that way we can create name that are not possible to be created in any other way. Like in definition of system processes.
https://github.com/rchain/rchain/blob/76f14c516bac3d6be2f91318c9b202b4239465bb/rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala#L76-L81

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/RhoRuntimeSyntax.scala [hunk 4]
```diff
@@ -22,24 +22,25 @@ final class RhoRuntimeOps[F[_]: Sync](
...
 
   def evaluate(
       term: String,
       initialPhlo: Cost,
-      normalizerEnv: Map[String, Par]
+      normalizerEnv: Map[String, Par],
+      randSeed: Blake2b512Random
   ): F[EvaluateResult] = {
-    implicit val rand: Blake2b512Random = Blake2b512Random(128)
+    implicit val rand: Blake2b512Random = randSeed
```
Comment (tgrospic) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/RhoRuntimeSyntax.scala]: Do we need this change here? We can make it on dev branch to reduce conflict on merging feature branch.

Comment (zsluedem) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/RhoRuntimeSyntax.scala]: https://github.com/rchain/rchain/blob/dev/crypto/src/main/scala/coop/rchain/crypto/hash/Blake2b512Random.scala#L137-L141
This is not a pure function. I can not predict unforgeable name in test without this change.

Comment (zsluedem) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/RhoRuntimeSyntax.scala]: Well. My bad. This change is not neccessary .

**Review by tgrospic — APPROVED** [review]
Nice work!

**Pull request merged**